From: Chad Monroe <chad@monroe.io>
Date: Fri, 07 Aug 2026 15:28:10 -0700
Subject: [PATCH] hostapd: synchronize bandwidth in AP/STA support

When an AP follows a co-located station (repeater), the station's
channel alone is not enough: the AP must also adopt its bandwidth.

wpa_supplicant reports ch_width, center_freq1 and bw320_offset from
the driver on channel switch events and in phy status. hostapd
passes them through apsta_state to freq_info, and uc_wpa_freq_info()
computes segment 0 and the center frequency for all widths,
including 20/40 MHz (USE_HT) and the second 320 MHz block.

A supplied ch_width takes precedence over scanning the generated
config for *_oper_chwidth. The fast path is guarded against a
missing value: ucode evaluates null >= 0 as true, and treating an
absent width as valid made every channel-only reload CSA drop to
20 MHz. uc_wpa_freq_info() rejects a missing width argument instead
of degrading to 20 MHz, bw320_offset falls back to 1 when absent or
invalid, and the CSA freq params are assigned on errno alone so
explicit zero values apply.

Based-on-work-by: Michael-CY Lee <michael-cy.lee@mediatek.com>
Signed-off-by: Chad Monroe <chad@monroe.io>
---
 openwrt/files/hostapd.uc        |   36 ++++++++----
 openwrt/files/wpa_supplicant.uc |    4 +
 src/ap/ucode.c                  |   47 +++++++++++++---
 src/utils/ucode.c               |   25 ++++++--
 wpa_supplicant/ucode.c          |   88 ++++++++++++++++++++++++------
 5 files changed, 159 insertions(+), 41 deletions(-)

--- a/src/ap/ucode.c
+++ b/src/ap/ucode.c
@@ -601,6 +601,9 @@ uc_hostapd_iface_stop(uc_vm_t *vm, size_
 	struct hostapd_iface *iface = uc_fn_thisval("hostapd.iface");
 	int i;
 
+	wpa_printf(MSG_INFO, "%s: stop iface for %s in state %s\n",
+			__func__, iface->phy, hostapd_state_text(iface->state));
+
 	if (!iface)
 		return NULL;
 
@@ -627,6 +630,9 @@ uc_hostapd_iface_start(uc_vm_t *vm, size
 	uint64_t intval;
 	int i;
 
+	wpa_printf(MSG_INFO, "%s: start iface for %s in state %s\n",
+			__func__, iface->phy, hostapd_state_text(iface->state));
+
 	if (!iface)
 		return NULL;
 
@@ -650,6 +656,7 @@ uc_hostapd_iface_start(uc_vm_t *vm, size
 	UPDATE_VAL(hw_mode, "hw_mode");
 	UPDATE_VAL(channel, "channel");
 	UPDATE_VAL(secondary_channel, "sec_channel");
+
 	if (!changed &&
 	    (iface->bss[0]->beacon_set_done ||
 	     iface->state == HAPD_IFACE_DFS))
@@ -700,6 +707,15 @@ out:
 		return ucv_boolean_new(true);
 	}
 
+	wpa_printf(MSG_INFO, "%s: updated channel=%d op_class=%d secondary channel=%d seg0=%d seg1=%d oper_chwidth=%d\n",
+			__func__,
+			conf->channel,
+			conf->op_class,
+			conf->secondary_channel,
+			hostapd_get_oper_centr_freq_seg0_idx(conf),
+			hostapd_get_oper_centr_freq_seg1_idx(conf),
+			hostapd_get_oper_chwidth(conf));
+
 	for (i = 0; i < iface->num_bss; i++) {
 		struct hostapd_data *hapd = iface->bss[i];
 		int ret;
@@ -734,13 +750,16 @@ uc_hostapd_iface_switch_channel(uc_vm_t
 	uint64_t intval;
 	int i, ret = 0;
 
+	wpa_printf(MSG_INFO, "%s: channel switch for %s\n", __func__, iface->phy);
 	if (!iface || ucv_type(info) != UC_OBJECT)
 		return NULL;
 
 	conf = iface->conf;
-	if ((intval = ucv_int64_get(ucv_object_get(info, "csa_count", NULL))) && !errno)
+	intval = ucv_int64_get(ucv_object_get(info, "csa_count", NULL));
+	if (!errno)
 		csa.cs_count = intval;
-	if ((intval = ucv_int64_get(ucv_object_get(info, "sec_channel", NULL))) && !errno)
+	intval = ucv_int64_get(ucv_object_get(info, "sec_channel", NULL));
+	if (!errno)
 		csa.freq_params.sec_channel_offset = intval;
 
 	csa.freq_params.ht_enabled = conf->ieee80211n;
@@ -769,13 +788,24 @@ uc_hostapd_iface_switch_channel(uc_vm_t
 		break;
 	}
 
-	if ((intval = ucv_int64_get(ucv_object_get(info, "frequency", NULL))) && !errno)
+	intval = ucv_int64_get(ucv_object_get(info, "frequency", NULL));
+	if (!errno)
 		csa.freq_params.freq = intval;
-	if ((intval = ucv_int64_get(ucv_object_get(info, "center_freq1", NULL))) && !errno)
+	intval = ucv_int64_get(ucv_object_get(info, "center_freq1", NULL));
+	if (!errno)
 		csa.freq_params.center_freq1 = intval;
-	if ((intval = ucv_int64_get(ucv_object_get(info, "center_freq2", NULL))) && !errno)
+	intval = ucv_int64_get(ucv_object_get(info, "center_freq2", NULL));
+	if (!errno)
 		csa.freq_params.center_freq2 = intval;
 
+	wpa_printf(MSG_INFO, "%s: CSA freq=%d bandwidth=%d sec_chan_offset=%d center_freq1=%d center_freq2=%d\n",
+			__func__,
+			csa.freq_params.freq,
+			csa.freq_params.bandwidth,
+			csa.freq_params.sec_channel_offset,
+			csa.freq_params.center_freq1,
+			csa.freq_params.center_freq2);
+
 	for (i = 0; i < iface->num_bss; i++)
 		ret = hostapd_switch_channel(iface->bss[i], &csa);
 
--- a/src/utils/ucode.c
+++ b/src/utils/ucode.c
@@ -105,6 +105,7 @@ uc_value_t *uc_wpa_freq_info(uc_vm_t *vm
 	uc_value_t *freq = uc_fn_arg(0);
 	uc_value_t *sec = uc_fn_arg(1);
 	int width = ucv_uint64_get(uc_fn_arg(2));
+	int bw320_offset = 1;
 	int freq_val, center_idx, center_ofs;
 	enum oper_chan_width chanwidth;
 	enum hostapd_hw_mode hw_mode;
@@ -113,6 +114,10 @@ uc_value_t *uc_wpa_freq_info(uc_vm_t *vm
 	int sec_channel = 0;
 	uc_value_t *ret;
 
+	/* a missing width argument must not degrade to 20 MHz */
+	if (errno)
+		return NULL;
+
 	if (ucv_type(freq) != UC_INTEGER)
 		return NULL;
 
@@ -145,6 +150,11 @@ uc_value_t *uc_wpa_freq_info(uc_vm_t *vm
 	case 9:
 		width = 3;
 		chanwidth = CONF_OPER_CHWIDTH_320MHZ;
+
+		/* bw320_offset is 1 for 320 MHz-1, and 2 for 320 MHz-2 */
+		bw320_offset = ucv_uint64_get(uc_fn_arg(3));
+		if (errno || (bw320_offset != 1 && bw320_offset != 2))
+			bw320_offset = 1;
 		break;
 	default:
 		return NULL;
@@ -176,12 +186,18 @@ uc_value_t *uc_wpa_freq_info(uc_vm_t *vm
 	ucv_object_add(ret, "hw_mode_str", ucv_string_new(modestr));
 	ucv_object_add(ret, "sec_channel", ucv_int64_new(sec_channel));
 	ucv_object_add(ret, "frequency", ucv_int64_new(freq_val));
+	ucv_object_add(ret, "oper_chwidth", ucv_int64_new(chanwidth));
 
-	if (!sec_channel)
+	if (chanwidth == CONF_OPER_CHWIDTH_USE_HT) {
+		center_idx = freq_val < 3000 ? 0 : channel + 2 * sec_channel;
+		ucv_object_add(ret, "center_seg0_idx", ucv_int64_new(center_idx));
+		ucv_object_add(ret, "center_freq1",
+			       ucv_int64_new(freq_val + sec_channel * 10));
 		return ret;
+	}
 
 	if (freq_val >= 5900)
-		center_ofs = 0;
+		center_ofs = 32 * (1 - bw320_offset);
 	else if (freq_val >= 5745)
 		center_ofs = 20;
 	else
@@ -190,10 +206,7 @@ uc_value_t *uc_wpa_freq_info(uc_vm_t *vm
 	tmp_channel &= ~((8 << width) - 1);
 	center_idx = tmp_channel + center_ofs + (4 << width) - 1;
 
-	if (freq_val < 3000)
-		ucv_object_add(ret, "center_seg0_idx", ucv_int64_new(0));
-	else
-		ucv_object_add(ret, "center_seg0_idx", ucv_int64_new(center_idx));
+	ucv_object_add(ret, "center_seg0_idx", ucv_int64_new(center_idx));
 	center_idx = (center_idx - channel) * 5 + freq_val;
 	ucv_object_add(ret, "center_freq1", ucv_int64_new(center_idx));
 
--- a/wpa_supplicant/ucode.c
+++ b/wpa_supplicant/ucode.c
@@ -10,7 +10,9 @@
 #include "config.h"
 #include "bss.h"
 #include "ucode.h"
+#include "driver_i.h"
 #include "offchannel.h"
+#include "common/ieee802_11_common.h"
 #ifdef CONFIG_DPP
 #include "common/dpp.h"
 #include "common/wpa_ctrl.h"
@@ -136,6 +138,8 @@ void wpas_ucode_event(struct wpa_supplic
 {
 	const char *state;
 	uc_value_t *val;
+	enum oper_chan_width ch_width;
+	int center_freq1, bw320_offset = 1;
 
 	if (event != EVENT_CH_SWITCH_STARTED)
 		return;
@@ -154,11 +158,42 @@ void wpas_ucode_event(struct wpa_supplic
 	uc_value_push(ucv_get(val));
 
 	if (event == EVENT_CH_SWITCH_STARTED) {
+		center_freq1 = data->ch_switch.cf1;
+
+		switch (data->ch_switch.ch_width) {
+		case CHAN_WIDTH_80:
+			ch_width = CONF_OPER_CHWIDTH_80MHZ;
+			break;
+		case CHAN_WIDTH_80P80:
+			ch_width = CONF_OPER_CHWIDTH_80P80MHZ;
+			break;
+		case CHAN_WIDTH_160:
+			ch_width = CONF_OPER_CHWIDTH_160MHZ;
+			break;
+		case CHAN_WIDTH_320:
+			ch_width = CONF_OPER_CHWIDTH_320MHZ;
+			break;
+		case CHAN_WIDTH_20_NOHT:
+		case CHAN_WIDTH_20:
+		case CHAN_WIDTH_40:
+		default:
+			ch_width = CONF_OPER_CHWIDTH_USE_HT;
+			break;
+		}
+
+		/* Check bandwidth 320 MHz-2 */
+		if (ch_width == CONF_OPER_CHWIDTH_320MHZ &&
+		    (center_freq1 == 6265) || center_freq1 == 6585 ||
+		     center_freq1 == 6905)
+			bw320_offset = 2;
+
 		ucv_object_add(val, "csa_count", ucv_int64_new(data->ch_switch.count));
 		ucv_object_add(val, "frequency", ucv_int64_new(data->ch_switch.freq));
 		ucv_object_add(val, "sec_chan_offset", ucv_int64_new(data->ch_switch.ch_offset));
-		ucv_object_add(val, "center_freq1", ucv_int64_new(data->ch_switch.cf1));
+		ucv_object_add(val, "center_freq1", ucv_int64_new(center_freq1));
 		ucv_object_add(val, "center_freq2", ucv_int64_new(data->ch_switch.cf2));
+		ucv_object_add(val, "ch_width", ucv_int64_new(ch_width));
+		ucv_object_add(val, "bw320_offset", ucv_int64_new(bw320_offset));
 	}
 
 	ucv_put(wpa_ucode_call(4));
@@ -567,26 +602,45 @@ out:
 }
 
 static void
-uc_wpas_iface_status_bss(uc_value_t *ret, struct wpa_bss *bss)
+uc_wpas_iface_status_bss(uc_value_t *ret, struct wpa_supplicant *wpa_s, struct wpa_bss *bss)
 {
 	int sec_chan = 0;
-	const u8 *ie;
-
-	ie = wpa_bss_get_ie(bss, WLAN_EID_HT_OPERATION);
-	if (ie && ie[1] >= 2) {
-		const struct ieee80211_ht_operation *ht_oper;
-		int sec;
-
-		ht_oper = (const void *) (ie + 2);
-		sec = ht_oper->ht_param & HT_INFO_HT_PARAM_SECONDARY_CHNL_OFF_MASK;
-		if (sec == HT_INFO_HT_PARAM_SECONDARY_CHNL_ABOVE)
-			sec_chan = 1;
-		else if (sec == HT_INFO_HT_PARAM_SECONDARY_CHNL_BELOW)
-			sec_chan = -1;
+	struct wpa_channel_info ci;
+	u8 op_class, channel;
+	enum oper_chan_width ch_width;
+	int center_freq1, bw320_offset = 1, is_24ghz;
+	enum hostapd_hw_mode hw_mode;
+
+	hw_mode = ieee80211_freq_to_chan(bss->freq, &channel);
+	is_24ghz = hw_mode == HOSTAPD_MODE_IEEE80211G ||
+		hw_mode == HOSTAPD_MODE_IEEE80211B;
+
+	wpa_drv_channel_info(wpa_s, &ci);
+	center_freq1 = ci.center_frq1;
+
+	if (bss->freq != center_freq1) {
+		if (is_24ghz)
+			sec_chan = (bss->freq < center_freq1) ? 1 : -1;
+		else
+			sec_chan = (bss->freq / 20) & 1 ? 1 : -1;
+	}
+
+	if (ieee80211_chaninfo_to_channel(ci.frequency, ci.chanwidth,
+					  sec_chan, &op_class, &channel))
+		return;
+
+	ch_width = op_class_to_ch_width(op_class);
+	if (ch_width == CONF_OPER_CHWIDTH_320MHZ &&
+		(center_freq1 == 6265) || center_freq1 == 6585 ||
+		 center_freq1 == 6905) {
+		/* Bandwidth 320 MHz-2 */
+		bw320_offset = 2;
 	}
 
 	ucv_object_add(ret, "sec_chan_offset", ucv_int64_new(sec_chan));
 	ucv_object_add(ret, "frequency", ucv_int64_new(bss->freq));
+	ucv_object_add(ret, "ch_width", ucv_int64_new(ch_width));
+	ucv_object_add(ret, "bw320_offset", ucv_int64_new(bw320_offset));
 }
 
 static void
@@ -622,7 +676,7 @@ uc_wpas_iface_status(uc_vm_t *vm, size_t
 
 	bss = wpa_s->current_bss;
 	if (bss)
-		uc_wpas_iface_status_bss(ret, bss);
+		uc_wpas_iface_status_bss(ret, wpa_s, bss);
 
 	if (wpa_s->valid_links) {
 		unsigned int valid_links = wpa_s->valid_links;
@@ -639,7 +693,7 @@ uc_wpas_iface_status(uc_vm_t *vm, size_t
 				continue;
 
 			link = ucv_object_new(vm);
-			uc_wpas_iface_status_bss(link, bss);
+			uc_wpas_iface_status_bss(link, wpa_s, bss);
 			ucv_array_set(links, i, link);
 		}
 
--- a/openwrt/files/hostapd.uc
+++ b/openwrt/files/hostapd.uc
@@ -112,6 +112,7 @@ start_disabled=1
 function iface_freq_info(iface, config, params)
 {
 	let freq = params.frequency;
+	let bw320_offset = params.bw320_offset;
 	if (!freq)
 		return null;
 
@@ -120,19 +121,23 @@ function iface_freq_info(iface, config,
 		sec_offset = 0;
 
 	let width = 0;
-	for (let line in config.radio.data) {
-		if (!sec_offset && match(line, /^ht_capab=.*HT40/)) {
-			sec_offset = null; // auto-detect
-			continue;
-		}
+	if (params.ch_width != null && params.ch_width >= 0) {
+		width = params.ch_width;
+	} else {
+		for (let line in config.radio.data) {
+			if (!sec_offset && match(line, /^ht_capab=.*HT40/)) {
+				sec_offset = null; // auto-detect
+				continue;
+			}
 
-		let val = match(line, /^(vht_oper_chwidth|he_oper_chwidth)=(\d+)/);
-		if (!val)
-			continue;
+			let val = match(line, /^(vht_oper_chwidth|he_oper_chwidth|eht_oper_chwidth)=(\d+)/);
+			if (!val)
+				continue;
 
-		val = int(val[2]);
-		if (val > width)
-			width = val;
+			val = int(val[2]);
+			if (val > width)
+				width = val;
+		}
 	}
 
 	if (freq < 4000)
@@ -147,7 +152,7 @@ function iface_freq_info(iface, config,
 	if (freq > 5925 && width > 0 && sec_offset == 0)
 		sec_offset = null;
 
-	return hostapd.freq_info(freq, sec_offset, width);
+	return hostapd.freq_info(freq, sec_offset, width, bw320_offset);
 }
 
 function iface_add(phy, config, phy_status, start_disabled)
@@ -1455,6 +1460,8 @@ let main_obj = {
 			up: true,
 			frequency: 0,
 			sec_chan_offset: 0,
+			ch_width: -1,
+			bw320_offset: 1,
 			csa: true,
 			csa_count: 0,
 			no_link: true,
@@ -1464,6 +1471,9 @@ let main_obj = {
 			if (req.args.up == null || !phy)
 				return libubus.STATUS_INVALID_ARGUMENT;
 
+			if (req.args.up)
+				hostapd.printf(`apsta_state update phy=${req.args.phy} freqeuncy=${req.args.frequency} sec_chan_offset=${req.args.sec_chan_offset} ch_width=${req.args.ch_width} bw320_offset=${req.args.bw320_offset} csa=${req.args.csa}`);
+
 			let config = hostapd.data.config[phy];
 			if (!config || !config.bss || !config.bss[0] || !config.bss[0].ifname)
 				return 0;
@@ -1507,6 +1517,8 @@ let main_obj = {
 				if (!freq_info)
 					return libubus.STATUS_UNKNOWN_ERROR;
 
+				hostapd.printf(`apsta_state phy=${phy} freq_info=${freq_info}`);
+
 				if (req.args.csa) {
 					freq_info.csa_count = req.args.csa_count ?? 10;
 					let ret = iface.switch_channel(freq_info);
--- a/openwrt/files/wpa_supplicant.uc
+++ b/openwrt/files/wpa_supplicant.uc
@@ -877,6 +877,8 @@ function iface_hostapd_notify(ifname, if
 		if (!mld) {
 			msg.frequency = status.frequency;
 			msg.sec_chan_offset = status.sec_chan_offset;
+			msg.ch_width = status.ch_width;
+			msg.bw320_offset = status.bw320_offset;
 		}
 		break;
 	default:
@@ -929,6 +931,8 @@ function iface_channel_switch(ifname, if
 		csa_count: info.csa_count ? info.csa_count - 1 : 0,
 		frequency: info.frequency,
 		sec_chan_offset: info.sec_chan_offset,
+		ch_width: info.ch_width,
+		bw320_offset: info.bw320_offset,
 	};
 
 	let mld = wpas.data.mld[ifname];
